boost: do not add the context-impl option for version below 1.65.0

mzeyen1985 opened the pull request:
https://github.com/spack/spack/pull/30654 doesn't prevent the addition of the b2 option `context-impl` when the variant is disabled.
This PR adds a version check to the corresponding if condition.

mzeyen1985 committed:
boost: do not access the context-impl variant for versions below 1.65.0

hainest commented:
Good catch! I really wish I had better CI for this package. This will be fixed in #30627, but we should merge this as a stop-gap.

tldahlgren commented on var/spack/repos/builtin/packages/boost/package.py:
```diff
@@ -489,7 +489,7 @@ def determine_b2_options(self, spec, options):
 
         # If we are building context, tell b2 which backend to use
-        if '+context' in spec:
+        if '+context' in spec and spec.version >= Version('1.65.0'):
```
Based on how versions are checked pretty much everywhere else in the package, I recommend using `spec.satisfies()` syntax for these changes for consistency.  See line 151 for example.

tldahlgren replied:
Actually, would this be better or simpler implemented as a conflict?

mzeyen1985 replied:
As far as I can tell, the problem is not `+context` but `context-impl` which are tied together.
Making the addition of context a conflict would also break the `coroutine` variant. See line 221 and 222.

Would the use of `spec.satisfies()` then look like the following?
```python
if '+context' in spec and spec.satisfies('@1.65.0:'):
```

tldahlgren replied:
Yes

mzeyen1985 replied:
Sorry for the late reply.
The latest change makes the fix agnostic of the boost version by simply checking if the `spec.variants` dict contains `context-impl`.

In summary:
- The `context-impl` variant is disabled if for boost versions < 1.65.0
- The content of the `context-impl` variant is only read if it is contained within `spec.variants`

hainest commented:
This PR does fix the problem, but we're silently hiding the issue. I agree with @tldahlgren, this should be made an explicit conflict.

```
$ spack install boost
...
==> boost: Successfully installed boost-1.79.0-adpxgb4djyo5jofn22zkbfjdgbyan2ro

$spack install boost+context
...
==> boost: Successfully installed boost-1.79.0-d623p6lo4fqo4cc6ntsnqq3wpkipq6n7

$ spack install boost~context
[+] ROOT/opt/spack/linux-ubuntu20.04-skylake/gcc-11.1.0/boost-1.79.0-adpxgb4djyo5jofn22zkbfjdgbyan2ro


$ spack install boost@1.64.0~context
...
==> boost: Successfully installed boost-1.64.0-wahwvdbc2zoybjrlfl52hs3vrc7f35kw

$ spack install boost@1.64.0+context
...
==> boost: Successfully installed boost-1.64.0-ideggoxhrecvdpz3njijbmqhpiisfhn4

$ spack install boost@1.64.0
[+] ROOT/opt/spack/linux-ubuntu20.04-skylake/gcc-11.1.0/boost-1.64.0-wahwvdbc2zoybjrlfl52hs3vrc7f35kw
```

mzeyen1985 committed:
boost: check if spec.variants contains context-impl

mzeyen1985 committed:
boost: improve error message when the context-impl variant causes a conflict

Executing spack solve boost@1.63.0 +context context-impl=fcontext
triggers the following error message:
==> Error: No version for 'boost' satisfies '@1.63.0' and '@1.79.0'

With this change, the error message becomes the following:
==> Error: Cannot set variant 'context-impl' for package 'boost' because the variant condition cannot be satisfied for the given spec

mzeyen1985 commented:
> > This PR does fix the problem, but we're silently hiding the issue. I agree with @tldahlgren, this should be made an explicit conflict.
> 
> @mzeyen1985 Did you want to add an explicit conflict to this PR?

cba17ad6b37fb6ba271669cb45fcf188717a5ede adds an explicit conflict for the `context-impl` variant improving the error message issued by spack.

Adding a conflict for `boost@:1.65.0 +context` as suggested above would be inconsistent with the following:
- https://github.com/spack/spack/blob/569884661992bd1dad39b1abb0b26167d6063baf/var/spack/repos/builtin/packages/boost/package.py#L219
- https://github.com/spack/spack/blob/f09f834a8fde3edad980b92de05b8b16b2dddada/var/spack/repos/builtin/packages/boost/package.py#L222
- https://github.com/spack/spack/blob/569884661992bd1dad39b1abb0b26167d6063baf/var/spack/repos/builtin/packages/boost/package.py#L225

tldahlgren commented:
@hainest @mzeyen1985 Is this PR ready to merge?

hainest commented:
@tldahlgren @mzeyen1985 This mostly solves the issue. I think the uncaught cases will be fixed by #30627. For example, it catches

    $ spack install boost@1.64.0 context-impl=fcontext
      ==> Error: Cannot set variant 'context-impl' for package 'boost' because the variant condition cannot be satisfied for the given spec

but not

    $ spack install boost@1.64.0+context   <-- This is broken, but not the subject of this PR
        ==> boost: Successfully installed boost-1.64.0-wgstburpeafwrqwptu5hztcqdqxcjaw7
        [+] ROOT/spack/opt/spack/linux-ubuntu20.04-skylake/gcc-11.1.0/boost-1.64.0-wgstburpeafwrqwptu5hztcqdqxcjaw7

    $ spack install boost@1.64.0 context-impl=fcontext
        [+] ROOT/spack/opt/spack/linux-ubuntu20.04-skylake/gcc-11.1.0/boost-1.64.0-wgstburpeafwrqwptu5hztcqdqxcjaw7

    $ spack/bin/spack install boost@1.64.0 context-impl=ucontext
        [+] ROOT/spack/opt/spack/linux-ubuntu20.04-skylake/gcc-11.1.0/boost-1.64.0-wgstburpeafwrqwptu5hztcqdqxcjaw7

I _think_ the reason here is that the concretizer is skipping the `context-impl` variant because the condition is `:1.65.0 +context` which isn't satisfied by the last two specs, even though `+context` was used to build the installed version. I will definitely add a test for this in my PR.
